### PR TITLE
Performance improvements for `chirho.robust`

### LINKED ISSUE #451: Potential slowdown of vmapping on the outside in `influence_fn`
In `ops.py`, 

```python
def influence_fn(...):
    ...
    @functools.wraps(target)
    def _fn(point: Point[T], *args: P.args, **kwargs: P.kwargs) -> S:
        param_eif = linearized(point, *args, **kwargs)
        return torch.func.jvp(
            lambda p: func_target(p, *args, **kwargs), (target_params,), (param_eif,)
        )[1]

    return _fn
```
Here if we vmap over `_fn`, then `func_target` keeps having to be evaluated. This might be a source of slowdown. One way to fix this would be to instead assume `linearized` batches over multiple datapoints and `_fn` takes as input multiple points instead of a single datapoint.

### PR DESCRIPTION
**Pull request by agrawalraj**
There are three main performance improvements:

1. Optimization when guide is a point estimate
If the guide is only a point estimate, using `NMCLogPredictiveLikelihood`, which relies on `torch.func.vmap` to batch over multiple datapoints can be orders of magnitude slower than manually batching the likelihood. In `test_empirical_fisher_vp_performance_with_likelihood` in `test_performance.py`,  we see that `NMCLogPredictiveLikelihood` can be 200x slower than manually vectorizing the likelihood using the new class `PointLogPredictiveLikelihood`.

2. Manually batching in `_flat_conjugate_gradient_solve`
One big performance hit was using `torch.func.vmap` to batch over multiple conjugate gradient solves. Since `torch.func.vmap` does not allow conditional if/else statements, the conjugate gradient solver cannot terminate earlier if the error tolerance is met. In this PR, we batch the conjugate gradient solver without using `torch.func.vmap` so that we can use conditional statements. In one experiment, I was able to run only 23 congugate gradient steps instead of a few hundred to reach error tolerance. This behavior is also exploited here: https://arxiv.org/abs/1903.08114. 

4. Batching over multiple points in `linearize` and `influence_fn`
As shown in  `test_performance.py`, simulating synthetic data from the model can be the most lengthy step. In the previous implementation, data is simulated each time the influence function is evaluated at a single point. In this PR, we evaluate the influence function over multiple points so that we only need to simulate data once.

Closes #451

### CODE REVIEW

**Review by eb8680 — COMMENTED** [review]
Nice detective work!

I can see why 2 and 3 would cause slowdowns, but 1 is still a bit mysterious to me even after spending some time yesterday playing around with the code and tests - there are a couple places where a naive implementation of `vmap` could introduce unnecessarily repeated computation, but I made some smaller changes that I thought would work around those and they didn't seem to make a difference.

At any rate, whatever is really going on, you've convinced me that it would be a good idea to replace all uses in `chirho.robust` of `torch.func.vmap` to batch over datapoints with the use of a manually vectorized version of `NMCLogPredictiveLikelihood` generalizing your `PointPredictiveLikelihood`. I'll create a separate issue for this and assign myself.

In the meantime, can we break this into smaller PRs that address the issues independently, starting with 2 and 3? I have some comments on API changes here but I think it would be easiest to address them in more isolated contexts.

File: tests/robust/test_performance.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+def test_empirical_fisher_vp_performance():
+    p = 500
+    loc = torch.tensor(1.0 * torch.ones(p), requires_grad=True)
+    cov_mat = torch.eye(p)
+    v = torch.ones(p)
+    func_log_prob = gaussian_log_prob
+    log_prob_params = {"loc": loc}
+    num_monte_carlo = 10000
+    start_time = time.time()
+    data = Predictive(GaussianModel(cov_mat), num_samples=num_monte_carlo)(loc)
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply tests/robust/test_performance.py:36]: Got it thanks!

File: tests/robust/test_performance.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    model_family, guide_family = model_guide
+
+    model = model_family(p=p)
+    guide = guide_family(p=p)
+
+    model()
+    guide()
+
+    start_time = time.time()
+    data = Predictive(
```
[the remark this answers is not in the document]
Comment (agrawalraj) [reply tests/robust/test_performance.py:110]: Makes sense thanks!

**Comment by agrawalraj** [comment]
> Nice detective work!
> 
> I can see why 2 and 3 would cause slowdowns, but 1 is still a bit mysterious to me even after spending some time yesterday playing around with the code and tests - there are a couple places where a naive implementation of `vmap` could introduce unnecessarily repeated computation, but I made some smaller changes that I thought would work around those and they didn't seem to make a difference.
> 
> At any rate, whatever is really going on, you've convinced me that it would be a good idea to replace all uses in `chirho.robust` of `torch.func.vmap` to batch over datapoints with the use of a manually vectorized version of `NMCLogPredictiveLikelihood` generalizing your `PointPredictiveLikelihood`. I'll create a separate issue for this and assign myself.
> 
> In the meantime, can we break this into smaller PRs that address the issues independently, starting with 2 and 3? I have some comments on API changes here but I think it would be easiest to address them in more isolated contexts.

Yup, I'll do that now!

**Comment by agrawalraj** [comment]
This PR will be broken up into 3 separate issues so will close it
